tensorflow_addons/optimizers/discriminative_layer_training.py

```diff
@@ -0,0 +1,254 @@
+
+        """Helper method to assign a layer's learning rate multiplier, which does nothing if lr mult is already set
+        """
+
+        try:
+            if layer.lr_mult and override:
+                layer.lr_mult = lr_mult  # check if layer has lr mult and if override, then assign the new lr mult
+        except AttributeError:
+            layer.lr_mult = lr_mult  # since layer has no lr mult, assign the mult
+
+    def _get_lowest_layers(self, layer, propagate_lr_mult_to_sub_layers=True):
+
+        """Helper method iterate through all nested layers of an object that behaves like a layer or model
+        By default, we want to propagate the lr mult to the lower layers.
+        tbh I can't properly explain how this works so see this post
+
+        https://stackoverflow.com/questions/6340351/iterating-through-list-of-list-in-python
+        """
+
+        mult = self._get_lr_mult(layer)
+        layers = self._get_layers(layer)
+
+        if len(layers) > 0:
+            for sublayer in layers:
+
+                # we generally want to propagate the lr mult to the lower layers
+                if propagate_lr_mult_to_sub_layers:
+                    self._assign_lr_mult(sublayer, mult)
+
+                # recursively iterate through the nested layers
+                for nested_sublayer in self._get_lowest_layers(sublayer):
+                    yield nested_sublayer
+
+        else:
+            yield layer
+
+    def _apply_lr_mult_to_var(self, layer):
+        """Helper method to apply the lr mult to the trainable variables of a layer
+        """
+
+        lr_mult = self._get_lr_mult(layer)
+
+        for var in layer.trainable_variables:
+            var.lr_mult = tf.convert_to_tensor(lr_mult, tf.float32)  # 0D tensor
+            var.lr_mult_value = (
+                lr_mult  # easier to check vars lr mult in graph and eager
+            )
+
+    def _check_for_lr_mult(self, layer, verbose=True, propagate=True):
+        """Identify which layers have an LR mult not equal to 1
+        """
+
+        layers_with_lr_mult = []
+
+        for sub_layer in self._get_lowest_layers(
+            layer, propagate_lr_mult_to_sub_layers=propagate
+        ):
+            lr_mult = self._get_lr_mult(sub_layer)
+            if lr_mult != 1.0:
+                layers_with_lr_mult.append(sub_layer)
+                if verbose:
+                    tf.print("layer %s lr_mult : %f" % (sub_layer.name, lr_mult))
+
+        return layers_with_lr_mult
+
+    def _compute_params(self, var_list):
+        """helps compute params to provide a summary that aligns with model.summary()
+        """
+        return np.sum([np.prod(list(var.shape)) for var in var_list])
+
+    def _prepare_model(self, model, verbose=True):
+        """Prepares a built model for disc training
+        """
+
+        layers_with_lr_mult = self._check_for_lr_mult(model, verbose=verbose)
+        if len(layers_with_lr_mult) == 0:
+
+            tf.print(
```

Comment:
Could you use logging to standardize with other modules. (info/warn/etc.):
https://github.com/tensorflow/addons/blob/cbb9fcc24853f0e0ba575e5b82f032c1ca14dc94/tensorflow_addons/layers/wrappers.py#L66

Reply:
will do